src/main/scala/com/databricks/spark/avro/AvroSaver.scala

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2014 Databricks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.databricks.spark.avro
+
+import java.sql.Timestamp
+import java.util.HashMap
+
+import scala.collection.immutable.Map
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql._
+import org.apache.spark.SparkContext._
+
+import org.apache.avro.generic.GenericData.Record
+import org.apache.avro.generic.GenericRecord
+import org.apache.avro.mapred._
+import org.apache.avro.{SchemaBuilder, Schema}
+
+import org.apache.hadoop.io.NullWritable
+import org.apache.hadoop.mapred.JobConf
+
+
+/** 
+ * This object provides a save() method that is used to save SchemaRDD as avro file.
+ * To do this, we first convert the schema and then convert each row of the RDD to corresponding
+ * avro types. One remark worth mentioning is the structName parameter that functions have. Avro
+ * records have a name associated with them, which must be unique. Since SturctType in sparkSQL
+ * doesn't have a name associated with it, we are taking the name of the last structure field that
+ * the current structure is a child of. For example if the row at the top level had a field called
+ * "X", which happens to be a structure, we would call that structure "X". When we process original
+ * rows, they get a name "topLevelRecord".
+ */
+object AvroSaver {
+
+  def save(schemaRDD: SchemaRDD, location: String): Unit = {
+    val jobConf = new JobConf(schemaRDD.sparkContext.hadoopConfiguration)
+    val builder = SchemaBuilder.record("topLevelRecord")
+    val schema = SchemaConverters.convertStructToAvro(schemaRDD.schema, builder)
+    AvroJob.setOutputSchema(jobConf, schema)
+
+    schemaRDD.mapPartitions(rowsToAvro(_, schemaRDD.schema)).saveAsHadoopFile(location,
+      classOf[AvroWrapper[GenericRecord]],
+      classOf[NullWritable],
+      classOf[AvroOutputFormat[GenericRecord]],
+      jobConf)
+  }
+
+  private def rowsToAvro(
+      rows: Iterator[Row],
+      schema: StructType): Iterator[(AvroKey[GenericRecord], NullWritable)] = {
+    val converter = createConverter(schema, "topLevelRecord")
+    val records = ArrayBuffer[(AvroKey[GenericRecord], NullWritable)]()
+
+    while (rows.hasNext) {
+      records.+=((new AvroKey(converter(rows.next).asInstanceOf[Record]), NullWritable.get()))
```

Comment:
Also I think it's reasonable to use `rows.map` here on the virtue of iterators being lazily computed and not requiring a whole partition to fit in memory (and writing an iterator from scratch is more code than it's worth to get around the Scala overhead).